File: app/src/main/java/com/qurantv/app/data/api/QuranComApi.kt
package com.qurantv.app.data.api

import kotlinx.serialization.json.Json

/** Quran.com API v4 (same Tanzil-sourced text; enrichment/fallback for D2). */
class QuranComApi(
    private val client: ApiClient,
    private val json: Json,
) {
    private val base = "https://api.quran.com/api/v4"

    suspend fun chapters(language: String): List<ChapterDto> =
        json.decodeFromString<ChaptersResponse>(client.getText("$base/chapters?language=$language")).chapters

    suspend fun versesUthmani(chapter: Int): List<VerseDto> =
        json.decodeFromString<VersesResponse>(client.getText("$base/quran/verses/uthmani?chapter_number=$chapter")).verses
}

File: web/src/data/api/QuranComApi.ts
// Ported 1:1 from app/src/main/java/com/qurantv/app/data/api/QuranComApi.kt
// Quran.com API v4 (same Tanzil-sourced text; enrichment/fallback for D2).

import type { ApiClient } from "./ApiClient";
import type { ChapterDto, VerseDto } from "./Dtos";

export class QuranComApi {
  private readonly base = "https://api.quran.com/api/v4";

  constructor(private readonly client: ApiClient) {}

  async chapters(language: string): Promise<ChapterDto[]> {
    const res = await this.client.getText(`${this.base}/chapters?language=${language}`);
    return (JSON.parse(res) as { chapters: ChapterDto[] }).chapters;
  }

  async versesUthmani(chapter: number): Promise<VerseDto[]> {
    const res = await this.client.getText(`${this.base}/quran/verses/uthmani?chapter_number=${chapter}`);
    return (JSON.parse(res) as { verses: VerseDto[] }).verses;
  }
}
